{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PROGETTO BIONFORMATICA DV \n",
    "------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Data Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import itertools \n",
    "import collections as cl\n",
    "import re\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15464246</td>\n",
       "      <td>15556529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000102010.15;Name=ENSG00000102010.15;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000130234.11;Name=ENSG00000130234.11;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561047</td>\n",
       "      <td>15675608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000285602.1;Name=ENSG00000285602.1;bi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  gene     start       end score strand frame  \\\n",
       "X    Ensembl  gene  15464246  15556529     .      +     .   \n",
       "X    Ensembl  gene  15561033  15602148     .      -     .   \n",
       "X    Ensembl  gene  15561047  15675608     .      -     .   \n",
       "\n",
       "                                          attributes  \n",
       "X  ID=ENSG00000102010.15;Name=ENSG00000102010.15;...  \n",
       "X  ID=ENSG00000130234.11;Name=ENSG00000130234.11;...  \n",
       "X  ID=ENSG00000285602.1;Name=ENSG00000285602.1;bi...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_path=\"https://raw.githubusercontent.com/LorenzoMauri/Progetto-Bioinformatica/master/data/ace2.gtf\"\n",
    "data= pd.read_csv(f_path,sep='\\t',skiprows=2,names=['chromosome','gene','start','end','score','strand','frame','attributes'])\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ID (stringa):\n",
    "    if re.findall('ID=(.*);',stringa) != None:\n",
    "        return \"\".join( re.findall('ID=(.*?);',stringa))\n",
    "    elif re.findall('ID=(.*)',stringa) != None:\n",
    "        return \"\".join( re.findall('ID=(.*?)',stringa))\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def nome (stringa):\n",
    "    if re.findall('Name=(.*);',stringa) != None:\n",
    "        return \"\".join( re.findall('Name=(.*?);',stringa))\n",
    "    elif re.findall('Name=(.*)',stringa) != None:\n",
    "        return \"\".join( re.findall('Name=(.*?)',stringa))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def biotype (stringa):\n",
    "    if re.findall('biotype=(.*);',stringa) !=[]:\n",
    "        return (\"\".join( re.findall('biotype=(.*);',stringa)))\n",
    "    elif re.findall('biotype=(.*)',stringa) != []:\n",
    "        return (\"\".join(  re.findall('biotype=(.*)',stringa)))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def parent (stringa):\n",
    "    if re.findall('Parent=(.*);',stringa) !=[]:\n",
    "        return(\"\".join(re.findall('Parent=(.*);',stringa)))\n",
    "    elif re.findall('Parent=(.*)',stringa) !=[]:\n",
    "        return(\"\".join(re.findall('Parent=(.*)',stringa)))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "data['ID']=data['attributes'].apply(ID)\n",
    "data['nome']=data['attributes'].apply(nome)\n",
    "data['biotype']=data['attributes'].apply(biotype)\n",
    "data['parent']=data['attributes'].apply(parent)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>nome</th>\n",
       "      <th>biotype</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15464246</td>\n",
       "      <td>15556529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000102010.15</td>\n",
       "      <td>ENSG00000102010.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561047</td>\n",
       "      <td>15675608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000285602.1</td>\n",
       "      <td>ENSG00000285602.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15602881</td>\n",
       "      <td>15621484</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000225833.1</td>\n",
       "      <td>ENSG00000225833.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15627318</td>\n",
       "      <td>15675012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000147003.7</td>\n",
       "      <td>ENSG00000147003.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  gene     start       end score strand frame                  ID  \\\n",
       "X    Ensembl  gene  15464246  15556529     .      +     .  ENSG00000102010.15   \n",
       "X    Ensembl  gene  15561033  15602148     .      -     .  ENSG00000130234.11   \n",
       "X    Ensembl  gene  15561047  15675608     .      -     .   ENSG00000285602.1   \n",
       "X    Ensembl  gene  15602881  15621484     .      +     .   ENSG00000225833.1   \n",
       "X    Ensembl  gene  15627318  15675012     .      -     .   ENSG00000147003.7   \n",
       "\n",
       "                 nome         biotype parent  \n",
       "X  ENSG00000102010.15  protein_coding   None  \n",
       "X  ENSG00000130234.11  protein_coding   None  \n",
       "X   ENSG00000285602.1  protein_coding   None  \n",
       "X   ENSG00000225833.1          lncRNA   None  \n",
       "X   ENSG00000147003.7  protein_coding   None  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=data.drop(columns='attributes')\n",
    "data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 . Leggere il file in un DataFrame, rimuovendo le righe che non si riferiscono al gene (il file contiene anche un miRNA).\n",
    "Chromosome X: 15561033-15602148 reverse strand (-)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>nome</th>\n",
       "      <th>biotype</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.5014.pooled.chrX.1</td>\n",
       "      <td>compmerge.5014.pooled.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1686.Testes.chrX.1</td>\n",
       "      <td>compmerge.1686.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.3044.pooled.chrX.1</td>\n",
       "      <td>compmerge.3044.pooled.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1209.Testes.chrX.1</td>\n",
       "      <td>compmerge.1209.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  gene     start       end score strand frame  \\\n",
       "X    Ensembl  gene  15561033  15602148     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600945     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600945     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600960     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600960     .      -     .   \n",
       "\n",
       "                             ID                          nome  \\\n",
       "X            ENSG00000130234.11            ENSG00000130234.11   \n",
       "X  compmerge.5014.pooled.chrX.1  compmerge.5014.pooled.chrX.1   \n",
       "X  compmerge.1686.Testes.chrX.1  compmerge.1686.Testes.chrX.1   \n",
       "X  compmerge.3044.pooled.chrX.1  compmerge.3044.pooled.chrX.1   \n",
       "X  compmerge.1209.Testes.chrX.1  compmerge.1209.Testes.chrX.1   \n",
       "\n",
       "              biotype parent  \n",
       "X      protein_coding   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto1=data[(data['start']>=15561033) &(data['end']<=15602148)]\n",
    "data_punto1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Calcolare il numero di esoni per ogni trascritto\n",
    "\n",
    "esoni con stesso parent e trascritti con stesso id "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(204, 22)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto1[(data_punto1['gene']=='exon')|(data_punto1['gene']=='transcript')]\n",
    "df_esoni = data_punto1[(data_punto1['gene']=='exon')]\n",
    "df_trascritti = data_punto1[(data_punto1['gene']=='transcript')]\n",
    "data_punto2=pd.merge(df_esoni,df_trascritti,how='inner',left_on='parent',right_on='ID')\n",
    "data_punto2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>gene_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>frame_x</th>\n",
       "      <th>ID_x</th>\n",
       "      <th>nome_x</th>\n",
       "      <th>biotype_x</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>score_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>frame_y</th>\n",
       "      <th>ID_y</th>\n",
       "      <th>nome_y</th>\n",
       "      <th>biotype_y</th>\n",
       "      <th>parent_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15600726</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00003897519</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15594845</td>\n",
       "      <td>15595003</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894057</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15592229</td>\n",
       "      <td>15592322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894058</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15591713</td>\n",
       "      <td>15591856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894059</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15589344</td>\n",
       "      <td>15589456</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894060</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_x gene_x   start_x     end_x score_x strand_x frame_x ID_x  \\\n",
       "0      Ensembl   exon  15600726  15600960       .        -       .        \n",
       "1      Ensembl   exon  15594845  15595003       .        -       .        \n",
       "2      Ensembl   exon  15592229  15592322       .        -       .        \n",
       "3      Ensembl   exon  15591713  15591856       .        -       .        \n",
       "4      Ensembl   exon  15589344  15589456       .        -       .        \n",
       "\n",
       "            nome_x biotype_x  ...      gene_y   start_y     end_y  score_y  \\\n",
       "0  ENSE00003897519      None  ...  transcript  15561033  15600960        .   \n",
       "1  ENSE00000894057      None  ...  transcript  15561033  15600960        .   \n",
       "2  ENSE00000894058      None  ...  transcript  15561033  15600960        .   \n",
       "3  ENSE00000894059      None  ...  transcript  15561033  15600960        .   \n",
       "4  ENSE00000894060      None  ...  transcript  15561033  15600960        .   \n",
       "\n",
       "   strand_y frame_y               ID_y             nome_y       biotype_y  \\\n",
       "0         -       .  ENST00000252519.8  ENST00000252519.8  protein_coding   \n",
       "1         -       .  ENST00000252519.8  ENST00000252519.8  protein_coding   \n",
       "2         -       .  ENST00000252519.8  ENST00000252519.8  protein_coding   \n",
       "3         -       .  ENST00000252519.8  ENST00000252519.8  protein_coding   \n",
       "4         -       .  ENST00000252519.8  ENST00000252519.8  protein_coding   \n",
       "\n",
       "             parent_y  \n",
       "0  ENSG00000130234.11  \n",
       "1  ENSG00000130234.11  \n",
       "2  ENSG00000130234.11  \n",
       "3  ENSG00000130234.11  \n",
       "4  ENSG00000130234.11  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto2.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Calcolare la lunghezza di ogni trascritto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_x', 'gene_x', 'start_x', 'end_x', 'score_x', 'strand_x',\n",
       "       'frame_x', 'ID_x', 'nome_x', 'biotype_x', 'parent_x', 'chromosome_y',\n",
       "       'gene_y', 'start_y', 'end_y', 'score_y', 'strand_y', 'frame_y', 'ID_y',\n",
       "       'nome_y', 'biotype_y', 'parent_y'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    39927\n",
       "1    39927\n",
       "2    39927\n",
       "3    39927\n",
       "4    39927\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lunghezza_trascritti = data_punto2['end_y']-data_punto2['start_y']\n",
    "lunghezza_trascritti.head(5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
