{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PROGETTO BIONFORMATICA DV \n",
    "------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Data Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import itertools \n",
    "import collections as cl\n",
    "import re\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15464246</td>\n",
       "      <td>15556529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000102010.15;Name=ENSG00000102010.15;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000130234.11;Name=ENSG00000130234.11;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561047</td>\n",
       "      <td>15675608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000285602.1;Name=ENSG00000285602.1;bi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   X chromosome feature     start       end score strand frame  \\\n",
       "0  X    Ensembl    gene  15464246  15556529     .      +     .   \n",
       "1  X    Ensembl    gene  15561033  15602148     .      -     .   \n",
       "2  X    Ensembl    gene  15561047  15675608     .      -     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000102010.15;Name=ENSG00000102010.15;...  \n",
       "1  ID=ENSG00000130234.11;Name=ENSG00000130234.11;...  \n",
       "2  ID=ENSG00000285602.1;Name=ENSG00000285602.1;bi...  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_path=\"https://raw.githubusercontent.com/LorenzoMauri/Progetto-Bioinformatica/master/data/ace2.gtf\"\n",
    "data= pd.read_csv(f_path,sep='\\t',skiprows=2,names=['X','chromosome','feature','start','end','score','strand','frame','attributes'])\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ID (stringa):\n",
    "    if re.findall('ID=(.*);',stringa) != None:\n",
    "        return \"\".join( re.findall('ID=(.*?);',stringa))\n",
    "    elif re.findall('ID=(.*)',stringa) != None:\n",
    "        return \"\".join( re.findall('ID=(.*?)',stringa))\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def nome (stringa):\n",
    "    if re.findall('Name=(.*);',stringa) != None:\n",
    "        return \"\".join( re.findall('Name=(.*?);',stringa))\n",
    "    elif re.findall('Name=(.*)',stringa) != None:\n",
    "        return \"\".join( re.findall('Name=(.*?)',stringa))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def biotype (stringa):\n",
    "    if re.findall('biotype=(.*);',stringa) !=[]:\n",
    "        return (\"\".join( re.findall('biotype=(.*);',stringa)))\n",
    "    elif re.findall('biotype=(.*)',stringa) != []:\n",
    "        return (\"\".join(  re.findall('biotype=(.*)',stringa)))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def parent (stringa):\n",
    "    if re.findall('Parent=(.*);',stringa) !=[]:\n",
    "        return(\"\".join(re.findall('Parent=(.*);',stringa)))\n",
    "    elif re.findall('Parent=(.*)',stringa) !=[]:\n",
    "        return(\"\".join(re.findall('Parent=(.*)',stringa)))\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "data['ID']=data['attributes'].apply(ID)\n",
    "data['nome']=data['attributes'].apply(nome)\n",
    "data['biotype']=data['attributes'].apply(biotype)\n",
    "data['parent']=data['attributes'].apply(parent)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "data=data.drop(columns='attributes')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 . Leggere il file in un DataFrame, rimuovendo le righe che non si riferiscono al gene (il file contiene anche un miRNA).\n",
    "Chromosome X: 15561033-15602148 reverse strand (-)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1524, 12)"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c=data.copy()\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene\n",
    "data.drop(c[(c['feature'] == 'gene') & (c['nome'] != 'ENSG00000130234.11')].index, inplace = True)\n",
    "\n",
    "\n",
    "# Trascritti\n",
    "data.drop(c[(c['feature'] == 'transcript') & (c['parent'] != 'ENSG00000130234.11')].index,inplace = True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>nome</th>\n",
       "      <th>biotype</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15464246</td>\n",
       "      <td>15556529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000102010.15</td>\n",
       "      <td>ENSG00000102010.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561047</td>\n",
       "      <td>15675608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000285602.1</td>\n",
       "      <td>ENSG00000285602.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15602881</td>\n",
       "      <td>15621484</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000225833.1</td>\n",
       "      <td>ENSG00000225833.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15627318</td>\n",
       "      <td>15675012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000147003.7</td>\n",
       "      <td>ENSG00000147003.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1519</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15664718</td>\n",
       "      <td>15664805</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSESTE00000059618</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSESTT00000015656.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1520</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15664337</td>\n",
       "      <td>15664395</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSESTE00000059619</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSESTT00000015656.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1521</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15659016</td>\n",
       "      <td>15659101</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSESTE00000059621</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSESTT00000015656.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1522</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15644916</td>\n",
       "      <td>15645029</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSESTE00000059609</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSESTT00000015656.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1523</td>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15639575</td>\n",
       "      <td>15639756</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSESTE00000059614</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSESTT00000015656.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1524 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      X chromosome feature     start       end score strand frame  \\\n",
       "0     X    Ensembl    gene  15464246  15556529     .      +     .   \n",
       "1     X    Ensembl    gene  15561033  15602148     .      -     .   \n",
       "2     X    Ensembl    gene  15561047  15675608     .      -     .   \n",
       "3     X    Ensembl    gene  15602881  15621484     .      +     .   \n",
       "4     X    Ensembl    gene  15627318  15675012     .      -     .   \n",
       "...  ..        ...     ...       ...       ...   ...    ...   ...   \n",
       "1519  X    Ensembl    exon  15664718  15664805     .      -     .   \n",
       "1520  X    Ensembl    exon  15664337  15664395     .      -     .   \n",
       "1521  X    Ensembl    exon  15659016  15659101     .      -     .   \n",
       "1522  X    Ensembl    exon  15644916  15645029     .      -     .   \n",
       "1523  X    Ensembl    exon  15639575  15639756     .      -     .   \n",
       "\n",
       "                      ID                nome         biotype  \\\n",
       "0     ENSG00000102010.15  ENSG00000102010.15  protein_coding   \n",
       "1     ENSG00000130234.11  ENSG00000130234.11  protein_coding   \n",
       "2      ENSG00000285602.1   ENSG00000285602.1  protein_coding   \n",
       "3      ENSG00000225833.1   ENSG00000225833.1          lncRNA   \n",
       "4      ENSG00000147003.7   ENSG00000147003.7  protein_coding   \n",
       "...                  ...                 ...             ...   \n",
       "1519                      ENSESTE00000059618            None   \n",
       "1520                      ENSESTE00000059619            None   \n",
       "1521                      ENSESTE00000059621            None   \n",
       "1522                      ENSESTE00000059609            None   \n",
       "1523                      ENSESTE00000059614            None   \n",
       "\n",
       "                    parent  \n",
       "0                     None  \n",
       "1                     None  \n",
       "2                     None  \n",
       "3                     None  \n",
       "4                     None  \n",
       "...                    ...  \n",
       "1519  ENSESTT00000015656.1  \n",
       "1520  ENSESTT00000015656.1  \n",
       "1521  ENSESTT00000015656.1  \n",
       "1522  ENSESTT00000015656.1  \n",
       "1523  ENSESTT00000015656.1  \n",
       "\n",
       "[1524 rows x 12 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>nome</th>\n",
       "      <th>biotype</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15602148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.5014.pooled.chrX.1</td>\n",
       "      <td>compmerge.5014.pooled.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1686.Testes.chrX.1</td>\n",
       "      <td>compmerge.1686.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.3044.pooled.chrX.1</td>\n",
       "      <td>compmerge.3044.pooled.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>X</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1209.Testes.chrX.1</td>\n",
       "      <td>compmerge.1209.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  gene     start       end score strand frame  \\\n",
       "X    Ensembl  gene  15561033  15602148     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600945     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600945     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600960     .      -     .   \n",
       "X    Ensembl  gene  15561033  15600960     .      -     .   \n",
       "\n",
       "                             ID                          nome  \\\n",
       "X            ENSG00000130234.11            ENSG00000130234.11   \n",
       "X  compmerge.5014.pooled.chrX.1  compmerge.5014.pooled.chrX.1   \n",
       "X  compmerge.1686.Testes.chrX.1  compmerge.1686.Testes.chrX.1   \n",
       "X  compmerge.3044.pooled.chrX.1  compmerge.3044.pooled.chrX.1   \n",
       "X  compmerge.1209.Testes.chrX.1  compmerge.1209.Testes.chrX.1   \n",
       "\n",
       "              biotype parent  \n",
       "X      protein_coding   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  \n",
       "X  aligned_transcript   None  "
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto1=data[(data['start']>=15561033)&(data['end']<=15602148)]\n",
    "data_punto1.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Calcolare il numero di esoni per ogni trascritto\n",
    "\n",
    "esoni con stesso parent e trascritti con stesso id "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(204, 22)"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto1[(data_punto1['gene']=='exon')|(data_punto1['gene']=='transcript')]\n",
    "df_esoni = data_punto1[(data_punto1['gene']=='exon')]\n",
    "df_trascritti = data_punto1[(data_punto1['gene']=='transcript')]\n",
    "data_punto2=pd.merge(df_esoni,df_trascritti,how='inner',left_on='parent',right_on='ID')\n",
    "data_punto2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "biotype\n",
       "aligned_transcript      12\n",
       "ccds_gene                1\n",
       "processed_transcript     3\n",
       "protein_coding           6\n",
       "dtype: int64"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto1.groupby(['biotype']).size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Calcolare la lunghezza di ogni trascritto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_x', 'gene_x', 'start_x', 'end_x', 'score_x', 'strand_x',\n",
       "       'frame_x', 'ID_x', 'nome_x', 'biotype_x', 'parent_x', 'chromosome_y',\n",
       "       'gene_y', 'start_y', 'end_y', 'score_y', 'strand_y', 'frame_y', 'ID_y',\n",
       "       'nome_y', 'biotype_y', 'parent_y'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto3=data_punto2\n",
    "data_punto2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    39927\n",
       "1    39927\n",
       "2    39927\n",
       "3    39927\n",
       "4    39927\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lunghezza_trascritti = data_punto2['end_y']-data_punto2['start_y']\n",
    "lunghezza_trascritti.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Calcolare le proteine associate a tali trascritti"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_y</th>\n",
       "      <th>biotype_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>165</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>166</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>167</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>168</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>169</td>\n",
       "      <td>transcript</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_y       biotype_y\n",
       "0    transcript  protein_coding\n",
       "1    transcript  protein_coding\n",
       "2    transcript  protein_coding\n",
       "3    transcript  protein_coding\n",
       "4    transcript  protein_coding\n",
       "..          ...             ...\n",
       "165  transcript  protein_coding\n",
       "166  transcript  protein_coding\n",
       "167  transcript  protein_coding\n",
       "168  transcript  protein_coding\n",
       "169  transcript  protein_coding\n",
       "\n",
       "[88 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto4=data_punto3\n",
    "data_punto3[(data_punto3['gene_y']=='transcript')&(data_punto3['biotype_y']=='protein_coding')][['gene_y','biotype_y']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Determinare le regioni genomiche (esone o porzione di esone) che sono presenti in tutti i trascritti. Ogni regione genomica è identificata da una posizione di inizio e una di fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_punto5=data_punto2\n",
    "#data_punto5[(data_punto5['start_y']<=data_punto5['start_x'])&(data_punto5['end_y']>=data_punto5['end_x'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>gene_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>frame_x</th>\n",
       "      <th>ID_x</th>\n",
       "      <th>nome_x</th>\n",
       "      <th>biotype_x</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>score_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>frame_y</th>\n",
       "      <th>ID_y</th>\n",
       "      <th>nome_y</th>\n",
       "      <th>biotype_y</th>\n",
       "      <th>parent_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15600726</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00003897519</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15594845</td>\n",
       "      <td>15595003</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894057</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15592229</td>\n",
       "      <td>15592322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894058</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15591713</td>\n",
       "      <td>15591856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894059</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15589344</td>\n",
       "      <td>15589456</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894060</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15571624</td>\n",
       "      <td>15571796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15570295</td>\n",
       "      <td>15570353</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>201</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15567726</td>\n",
       "      <td>15567826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>202</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15566253</td>\n",
       "      <td>15566369</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>203</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15564218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome_x gene_x   start_x     end_x score_x strand_x frame_x ID_x  \\\n",
       "0        Ensembl   exon  15600726  15600960       .        -       .        \n",
       "1        Ensembl   exon  15594845  15595003       .        -       .        \n",
       "2        Ensembl   exon  15592229  15592322       .        -       .        \n",
       "3        Ensembl   exon  15591713  15591856       .        -       .        \n",
       "4        Ensembl   exon  15589344  15589456       .        -       .        \n",
       "..           ...    ...       ...       ...     ...      ...     ...  ...   \n",
       "199      Ensembl   exon  15571624  15571796       .        -       .        \n",
       "200      Ensembl   exon  15570295  15570353       .        -       .        \n",
       "201      Ensembl   exon  15567726  15567826       .        -       .        \n",
       "202      Ensembl   exon  15566253  15566369       .        -       .        \n",
       "203      Ensembl   exon  15564024  15564218       .        -       .        \n",
       "\n",
       "              nome_x biotype_x  ...      gene_y   start_y     end_y  score_y  \\\n",
       "0    ENSE00003897519      None  ...  transcript  15561033  15600960        .   \n",
       "1    ENSE00000894057      None  ...  transcript  15561033  15600960        .   \n",
       "2    ENSE00000894058      None  ...  transcript  15561033  15600960        .   \n",
       "3    ENSE00000894059      None  ...  transcript  15561033  15600960        .   \n",
       "4    ENSE00000894060      None  ...  transcript  15561033  15600960        .   \n",
       "..               ...       ...  ...         ...       ...       ...      ...   \n",
       "199                       None  ...  transcript  15564024  15600960        .   \n",
       "200                       None  ...  transcript  15564024  15600960        .   \n",
       "201                       None  ...  transcript  15564024  15600960        .   \n",
       "202                       None  ...  transcript  15564024  15600960        .   \n",
       "203                       None  ...  transcript  15564024  15600960        .   \n",
       "\n",
       "     strand_y frame_y                          ID_y  \\\n",
       "0           -       .             ENST00000252519.8   \n",
       "1           -       .             ENST00000252519.8   \n",
       "2           -       .             ENST00000252519.8   \n",
       "3           -       .             ENST00000252519.8   \n",
       "4           -       .             ENST00000252519.8   \n",
       "..        ...     ...                           ...   \n",
       "199         -       .  compmerge.1687.Testes.chrX.1   \n",
       "200         -       .  compmerge.1687.Testes.chrX.1   \n",
       "201         -       .  compmerge.1687.Testes.chrX.1   \n",
       "202         -       .  compmerge.1687.Testes.chrX.1   \n",
       "203         -       .  compmerge.1687.Testes.chrX.1   \n",
       "\n",
       "                           nome_y           biotype_y  \\\n",
       "0               ENST00000252519.8      protein_coding   \n",
       "1               ENST00000252519.8      protein_coding   \n",
       "2               ENST00000252519.8      protein_coding   \n",
       "3               ENST00000252519.8      protein_coding   \n",
       "4               ENST00000252519.8      protein_coding   \n",
       "..                            ...                 ...   \n",
       "199  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "200  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "201  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "202  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "203  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "\n",
       "                         parent_y  \n",
       "0              ENSG00000130234.11  \n",
       "1              ENSG00000130234.11  \n",
       "2              ENSG00000130234.11  \n",
       "3              ENSG00000130234.11  \n",
       "4              ENSG00000130234.11  \n",
       "..                            ...  \n",
       "199  compmerge.1687.Testes.chrX.1  \n",
       "200  compmerge.1687.Testes.chrX.1  \n",
       "201  compmerge.1687.Testes.chrX.1  \n",
       "202  compmerge.1687.Testes.chrX.1  \n",
       "203  compmerge.1687.Testes.chrX.1  \n",
       "\n",
       "[204 rows x 22 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15561033"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto5['start_y'].min()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15602148"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto5['end_y'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_x</th>\n",
       "      <th>gene_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>frame_x</th>\n",
       "      <th>ID_x</th>\n",
       "      <th>nome_x</th>\n",
       "      <th>biotype_x</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>score_y</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>frame_y</th>\n",
       "      <th>ID_y</th>\n",
       "      <th>nome_y</th>\n",
       "      <th>biotype_y</th>\n",
       "      <th>parent_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15600726</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00003897519</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15594845</td>\n",
       "      <td>15595003</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894057</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15592229</td>\n",
       "      <td>15592322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894058</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15591713</td>\n",
       "      <td>15591856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894059</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15589344</td>\n",
       "      <td>15589456</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00000894060</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15561033</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>ENST00000252519.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000130234.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15571624</td>\n",
       "      <td>15571796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15570295</td>\n",
       "      <td>15570353</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>201</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15567726</td>\n",
       "      <td>15567826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>202</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15566253</td>\n",
       "      <td>15566369</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>203</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15564218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15564024</td>\n",
       "      <td>15600960</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "      <td>aligned_transcript</td>\n",
       "      <td>compmerge.1687.Testes.chrX.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome_x gene_x   start_x     end_x score_x strand_x frame_x ID_x  \\\n",
       "0        Ensembl   exon  15600726  15600960       .        -       .        \n",
       "1        Ensembl   exon  15594845  15595003       .        -       .        \n",
       "2        Ensembl   exon  15592229  15592322       .        -       .        \n",
       "3        Ensembl   exon  15591713  15591856       .        -       .        \n",
       "4        Ensembl   exon  15589344  15589456       .        -       .        \n",
       "..           ...    ...       ...       ...     ...      ...     ...  ...   \n",
       "199      Ensembl   exon  15571624  15571796       .        -       .        \n",
       "200      Ensembl   exon  15570295  15570353       .        -       .        \n",
       "201      Ensembl   exon  15567726  15567826       .        -       .        \n",
       "202      Ensembl   exon  15566253  15566369       .        -       .        \n",
       "203      Ensembl   exon  15564024  15564218       .        -       .        \n",
       "\n",
       "              nome_x biotype_x  ...      gene_y   start_y     end_y  score_y  \\\n",
       "0    ENSE00003897519      None  ...  transcript  15561033  15600960        .   \n",
       "1    ENSE00000894057      None  ...  transcript  15561033  15600960        .   \n",
       "2    ENSE00000894058      None  ...  transcript  15561033  15600960        .   \n",
       "3    ENSE00000894059      None  ...  transcript  15561033  15600960        .   \n",
       "4    ENSE00000894060      None  ...  transcript  15561033  15600960        .   \n",
       "..               ...       ...  ...         ...       ...       ...      ...   \n",
       "199                       None  ...  transcript  15564024  15600960        .   \n",
       "200                       None  ...  transcript  15564024  15600960        .   \n",
       "201                       None  ...  transcript  15564024  15600960        .   \n",
       "202                       None  ...  transcript  15564024  15600960        .   \n",
       "203                       None  ...  transcript  15564024  15600960        .   \n",
       "\n",
       "     strand_y frame_y                          ID_y  \\\n",
       "0           -       .             ENST00000252519.8   \n",
       "1           -       .             ENST00000252519.8   \n",
       "2           -       .             ENST00000252519.8   \n",
       "3           -       .             ENST00000252519.8   \n",
       "4           -       .             ENST00000252519.8   \n",
       "..        ...     ...                           ...   \n",
       "199         -       .  compmerge.1687.Testes.chrX.1   \n",
       "200         -       .  compmerge.1687.Testes.chrX.1   \n",
       "201         -       .  compmerge.1687.Testes.chrX.1   \n",
       "202         -       .  compmerge.1687.Testes.chrX.1   \n",
       "203         -       .  compmerge.1687.Testes.chrX.1   \n",
       "\n",
       "                           nome_y           biotype_y  \\\n",
       "0               ENST00000252519.8      protein_coding   \n",
       "1               ENST00000252519.8      protein_coding   \n",
       "2               ENST00000252519.8      protein_coding   \n",
       "3               ENST00000252519.8      protein_coding   \n",
       "4               ENST00000252519.8      protein_coding   \n",
       "..                            ...                 ...   \n",
       "199  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "200  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "201  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "202  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "203  compmerge.1687.Testes.chrX.1  aligned_transcript   \n",
       "\n",
       "                         parent_y  \n",
       "0              ENSG00000130234.11  \n",
       "1              ENSG00000130234.11  \n",
       "2              ENSG00000130234.11  \n",
       "3              ENSG00000130234.11  \n",
       "4              ENSG00000130234.11  \n",
       "..                            ...  \n",
       "199  compmerge.1687.Testes.chrX.1  \n",
       "200  compmerge.1687.Testes.chrX.1  \n",
       "201  compmerge.1687.Testes.chrX.1  \n",
       "202  compmerge.1687.Testes.chrX.1  \n",
       "203  compmerge.1687.Testes.chrX.1  \n",
       "\n",
       "[204 rows x 22 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_punto5[(data_punto5['start_x']>=data_punto5['start_y'].min())&(data_punto5['end_x']<=data_punto5['end_y'].max())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'data_punto3' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-166-a43c149ca080>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mdata_punto3\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'data_punto3' is not defined"
     ]
    }
   ],
   "source": [
    "data_punto3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
